src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -4110,26 +4111,43 @@ void MergeTreeData::backupData(BackupEntriesCollector & backup_entries_collector
     backup_entries_collector.addBackupEntries(backupParts(data_parts, data_path_in_backup));
 }
 
-BackupEntries MergeTreeData::backupParts(const DataPartsVector & data_parts, const String & data_path_in_backup)
+BackupEntries MergeTreeData::backupParts(const DataPartsVector & data_parts, const String & data_path_in_backup) const
 {
     BackupEntries backup_entries;
     std::map<DiskPtr, std::shared_ptr<TemporaryFileOnDisk>> temp_dirs;
 
+    /// Tables in atomic databases have UUID. When using atomic database we don't have to create hard links to make a backup, we can just
```

Comment:
Database engine and data parts cleanup mechanism are two completely orthogonal things. Outdated data parts are removed exactly the same way in all database engines (Ordinary, Atomic, Replicated, even in Memory and MaterializedMySQL), because `MergeTreeData` is responsible for this, not database engine. It's always enough to hold `DataPartPtr`s to prevent parts from being removed. 

However, you also need to prevent the whole table from being dropped and that's where database engine matters. For tables without UUID you need to hold shared table lock (see `IStorage::lockForShare`) and for tables without UUID you need to hold `StoragePtr`.

Reply:
As I already said:
> to prevent the whole table from being dropped

More precisely, to protect `IStorage` from calling `IStorage::drop` and from doing something like `rm -rf` for the whole table data directory. See `InterpreterDropQuery`, `IDatabase::dropTable` and `DatabaseCatalog::enqueueDroppedTableCleanup` for details.

Reply:
I changed my implementation, now it either holds `StoragePtr + DataPartPtr` or holds `StoragePtr + DataPartPtr + lockForShare` or uses hard links depenending on the situation.